skorch/helper.py

```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
+      The dtype to cast the categorical values to. If None, don't
+      change them. If you do this, it can happen that the categorical
+      values will have different dtypes, reflecting the number of
+      unique categories.
+
+    Notes
+    -----
+    The value of X will always be 2-dimensional, even if it only
+    contains 1 column.
+
+    """
+    import pandas as pd
+
+    def __init__(
+            self,
+            treat_int_as_categorical=False,
+            float_dtype=np.float32,
+            int_dtype=np.int64,
+    ):
+        self.treat_int_as_categorical = treat_int_as_categorical
+        self.float_dtype = float_dtype
+        self.int_dtype = int_dtype
+
+    def _check_dtypes(self, df):
+        """Perform a check on the DataFrame to detect wrong dtypes or keys.
+
+        Makes sure that there are no conflicts in key names.
+
+        If dtypes are found that cannot be dealt with, raises a
+        TypeError with a message indicating which ones caused trouble.
+
+        Raises
+        ------
+        ValueError
+          If there already is a column named 'X'.
+
+        TypeError
+          If a wrong dtype is found.
+
+        """
+        if 'X' in df:
+            raise ValueError(
+                "DataFrame contains a column named 'X', which clashes "
+                "with the name chosen for cardinal features; consider "
+                "renaming that column.")
+
+        wrong_dtypes = []
+
+        for col, dtype in zip(df, df.dtypes):
+            if isinstance(dtype, self.pd.core.dtypes.dtypes.CategoricalDtype):
+                continue
+            if np.issubdtype(dtype, np.integer):
+                continue
+            if np.issubdtype(dtype, np.floating):
+                continue
+            wrong_dtypes.append((col, dtype))
+
+        if not wrong_dtypes:
+            return
+
+        wrong_dtypes = sorted(wrong_dtypes, key=lambda tup: tup[0])
+        msg_dtypes = ", ".join(
+            "{} ({})".format(col, dtype) for col, dtype in wrong_dtypes)
+        msg = ("The following columns have dtypes that cannot be "
+               "interpreted as numerical dtypes: {}".format(msg_dtypes))
+        raise TypeError(msg)
+
+    # pylint: disable=unused-argument
+    def fit(self, df, y=None, **fit_params):
+        self._check_dtypes(df)
+        return self
+
+    def transform(self, df):
+        """Transform DataFrame to become a dict that works well with skorch.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+          Incoming DataFrame.
+
+        Returns
+        -------
+        X_dict: dict
+          Dictionary with all floats concatenated using the key "X"
+          and all categorical values encoded as integers, using their
+          respective column names as keys.
+
+        """
+        self._check_dtypes(df)
+
+        X_dict = {}
+        Xf = []  # floats
+
+        for col, dtype in zip(df, df.dtypes):
+            values = df[col]
```

Comment:
Nit: Having `values.values` later is a bit harder to follow. Maybe `X_col=df[col`?

Reply:
Not a nit :+1: